language-support/ts/daml-ledger/index.ts

```diff
@@ -552,33 +553,53 @@ class Ledger {
         console.error(`Ledger.${callerName} unknown message`, json);
       }
     };
-    const onClose = ({ code, reason }: { code: number; reason: string }): void => {
-      emitter.emit('close', {code, reason});
-      const now = new Date().getTime();
-      // we try to reconnect if we could connect previously and we were live for at least
-      // 'reconnectThreshold'.
-      if (lastOffset !== undefined && isLiveSince !== undefined && now - isLiveSince >= this.reconnectThreshold) {
-        isLiveSince = undefined;
-        isReconnecting = true;
-        ws = new WebSocket(this.wsBaseUrl + endpoint, protocols);
-        ws.addEventListener('open', onOpen);
-        ws.addEventListener('message', onMessage);
-        ws.addEventListener('close', onClose);
-      }
+    const closeStream = (status: { code: number; reason: string }): void => {
+      streamClosed = true;
+      emitter.emit('close', status);
+      emitter.removeAllListeners();
+    };
+    const onWsClose = (): void => {
+      if (streamClosed === false) {
+        const now = new Date().getTime();
+        // we want to try and keep the stream open, so we try to reconnect
+        // the underlying ws
+        if (lastOffset !== undefined && isLiveSince !== undefined && now - isLiveSince >= this.reconnectThreshold) {
+          isLiveSince = undefined;
+          isReconnecting = true;
+          ws = new WebSocket(this.wsBaseUrl + endpoint, protocols);
+          ws.addEventListener('open', onWsOpen);
+          ws.addEventListener('message', onWsMessage);
+          ws.addEventListener('close', onWsClose);
+        } else {
+          // ws has closed too quickly / never managed to connect: we give up
+          closeStream({code: 4001, reason: 'ws connection failed'});
```

Comment:
Is there a reason why you chose 4001 and 4000 or are those somewhat arbitrary choices?

Reply:
Somewhat arbitrary, but <4000 is reserved for the protocol, 4000+ is for applications.